### ISSUE TITLE
could not find a text editor to use (when not specifying one)

### ISSUE THREAD

**Issue by balthisar** [piotrmurach/tty-editor#3]
### Are you in the right place?
* For issues or feature requests file a GitHub issue in this repository
* For general questions or discussion post in [Gitter](https://gitter.im/piotrmurach/tty)

### Describe the problem

`TTY::Editor.new()` fails with _could not find a text editor to use. Please specify $VISUAL or $EDITOR or install one of the following editors: nano, notepad, vim, vi, emacs, code, subl, mate, atom, pico, qe, mg, jed._

### Steps to reproduce the problem
```
      puts ENV["EDITOR"]   # prints /usr/local/bin/bbedit 
      editor = TTY::Editor.new() # causes TTY::Editor::EditorNotFoundError
```

### Actual behaviour

`TTY::Editor.new()` fails with _could not find a text editor to use. Please specify $VISUAL or $EDITOR or install one of the following editors: nano, notepad, vim, vi, emacs, code, subl, mate, atom, pico, qe, mg, jed._

### Expected behaviour
I expect the editor instance to be created without error when use with my default text editor.
Note that is I specify an editor specifically, e.g., `TTY::Editor.new(command:"bbedit")` everything works. However I'm looking for cross-platform, editor-neutral behavior, this using this gem.

### Describe your environment

* OS version: macOS 11.3.1
* Ruby version: ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin20]
* TTY::Editor version: 0.6.0

**Comment by piotrmurach** [comment]
Hi Jim 👋 

Thanks for using `tty-editor` and reporting this issue.

You're right this is meant to be a cross-platform way to invoke an editor. Let's debug the situation to see what is going on. Would you mind providing me with some answers from your system?

As a priority by default the `$VISUAL` and `$EDITOR` variables are checked. So could you please check what you get when you run the following in your shell:

```shell
echo $VISUAL #  => ???
echo $EDITOR # => ???
```

Also, `tty-editor` checks if a command can be run by looking for it in the path directories. Let's see if `/usr/local/bin` is included in your system:

```shell
echo $PATH | grep -oe /usr/local/bin # => ???
```

Then in IRB console:

```ruby
require "tty-editor"
TTY::Editor.from_env # => ????
TTY::Editor.available # => ???
TTY::Editor.exist?("bbedit") # => ???
```

**Comment by balthisar** [comment]
Thanks. In my shell, both `$VISUAL` and `$EDITOR` return the same as in the bug report: `/usr/local/bin/bbedit`

Performing `echo $PATH | grep -oe /usr/local/bin` results in `/usr/local/bin`, which I would expect, because I have my bbedit and many, many other executables there.

From IRB:

```
TY::Editor.from_env
 => ["/usr/local/bin/bbedit", "/usr/local/bin/bbedit"]
```
```
TTY::Editor.available
 => [] 
```

```TTY::Editor.exist?("bbedit")
 => true 
```

Now, I've not had time to step through the program logic yet. I assumed that my $EDITOR would be trusted, especially if it leads to an executable. I see the const EXECUTABLES, so I'm not sure if there's a check to ensure that my editor is in the array (it isn't). As I said, if I use command: bbedit, it works. And ironically, it works flawlessly in my Windows VM, including giving me the choice between Notepad and the Visual Code editor (it would be nice to see Notepad++ as an option, but on the other hand I don't really give a crap about Windows).

And just to confirm, if I set my $EDITOR to, say, `/usr/bin/nano`, it works perfectly fine. And nano is in the EXECUTABLES const array.

**Comment by piotrmurach** [comment]
Thanks for providing the answers, they helped a lot. I know exactly what the issue is. It's to do with a method that checks if a command exists on the system, it assumes that commands are given as names and not absolute paths. In the past, this gem relied on my other gem [tty-which](https://github.com/piotrmurach/tty-which) to handle this part. I decided to replace it with a custom method that introduced a regression that fails when the executable is given as an absolute path. The `nano` example works, but not because it is found using the `$EDITOR` value but because it's in a default list of editors(the very first) that this gem checks and hence is correctly identified as an 'existing' editor. I shall fix this problem.

As for Windows, I thought about adding `notepad++` but usually, this editor's executable is not in the user's path, at least that's the case in my Windows VM. However, I see nothing against adding it in cases where a user has correctly set it up(obviously not me). Would you mind checking if `notepad++` runs fine when you configure it with `tty-editor` as the main command on Windows?

**Comment by balthisar** [comment]
I've just verified with `notepad++` on Windows, but it's probably not very representative of most installations. Yes, `command: 'notepad++'` works, but I've installed it via a Windows package manager called "chocolately" that installs things in a path that's in Windows' `$PATH`. I suspect that if I'd installed the portable version (which I normally do), it wouldn't be in the path, and would fail.

FYI just as a datapoint: if I `set EDITOR=code` in Windows, then `tty-editor` is smart enough not to ask me to choose, which is nice. Apparently Windows doesn't set `%EDITOR%` or `%VISUAL%` by default (and has ugly references to variables, to boot).

Based on that, I tried `unset EDITOR` on my Mac, but I get the same failure as the original bug report: `Please specify $VISUAL or $EDITOR or install one of the following editors: …` If I also `unset VISUAL`, then I get to choose from one of the built-in editors, like on Windows.

I wonder if `tty-editor` should fall back to this behaviour automatically when the editor in EDITOR or VISUAL can't be found or used?

Anyway, thanks for the help!

**Comment by piotrmurach** [comment]
Currently, when `EDITOR` and `VISUAL` have non-empty values, these are only checked and if executables they refer to are not available the search stops and an error gets raised. You're right, the check should continue and search through all default editors list before giving up. That's another issue that needs to be improved. Thanks!

**Comment by balthisar** [comment]
No issue, my temporary workaround works, and I understand what a really, really big backlog looks like! Cheers!

**Comment by piotrmurach** [comment]
@balthisar Both issues should be fixed now. Would you have time to check against the master branch that fixes are working fine? I have everything wrapped up to make a new release.

**Comment by piotrmurach** [comment]
@balthisar Thanks. I added [default editors list](https://github.com/piotrmurach/tty-editor#3-default-editors) to the readme. I was thinking of including `gedit` and `kate` editors before making the release. I think they could be good additions to the list based on their popularity. I'm trying to keep the list concise to include the most popular text editors. Any thoughts?

**Comment by balthisar** [comment]
I'm not really sure who else is using the tty gems, meaning I don't know the audience of the people using the gems using your gems, further meaning I don't know what their normal user expectations are when working in a terminal.

Usually things just work when I set `$EDITOR`, but git is a nasty exception that doesn't observe my setting. It has `core.editor`, though.  For people who live constantly in the terminal, this is probably enough, and 0.7 now works the way I think most people expect. I love the idea of the menu, though, for people that might not spend a lot of time in the terminal, letting them choose what they want instead of having to Google how to exit vim when they unexpectedly get there!

With that, I think `gedit` is gnome's default, so that's entirely reasonable to add to the list. I don't know a thing about `kate`.

**Comment by piotrmurach** [comment]
Thanks for your thoughts. The audience for tty components is wide. However, I try to make choices that are the most flexible going forward and make the components most reusable in many contexts. 

The `kate` is apparently a default KDE editor which would help cover a wide spectrum of most popular choices when the default editor isn't configured or found. This is a balancing act of keeping the list of editors small and still useful. 

I'm thinking of sorting default editors in alphabetical order. Currently, they are displayed and checked in the 'most likely present' order. Can you see any problem with this change?

**Comment by balthisar** [comment]
I don't see any problems, but if you're looking for input, if they're not present, they won't be displayed as options, right? In which case I'd go for alphabetical. End users (not the folks implementing your gem) won't have any idea why any other order exists, and are likely to think that the list isn't sorted at all.

**Comment by piotrmurach** [comment]
Correct, the menu will only show editors that are available on the user's system. The menu gets also triggered when, for example, more than one editor is provided with the `:command` keyword option or `EDITOR` and `VISUAL` are not configured. 

I changed the default list to be sorted. An ordered list may help in situation when the majority of default editors are present. These will be paged, so it may help in searching through the list for your favourite text editor.

I released `0.7.0` with all the changes. Thank you for answering my questions and contributing suggestions that made this release more robust!
